compiler/p/codegen/OMRMemoryReference.cpp

```diff
@@ -1677,56 +1673,54 @@ void OMR::Power::MemoryReference::accessStaticItem(TR::Node *node, TR::SymbolRef
-            loadAddressConstant(cg, nodeForSymbol, TR_CountForRecompile, reg, NULL, false, TR_GlobalValue);
-            return;
-            }
-         else if (symbol->isRecompilationCounter())
-            {
-            TR::Register *reg = _baseRegister = cg->allocateRegister();
-            loadAddressConstant(cg, nodeForSymbol, 0, reg, NULL, false, TR_BodyInfoAddressLoad);
-            return;
-            }
-         else if (symbol->isCompiledMethod())
-            {
-            TR::Register *reg = _baseRegister = cg->allocateRegister();
-            loadAddressConstant(cg, nodeForSymbol, 0, reg, NULL, false, TR_RamMethodSequence);
-            return;
-            }
-         else if (symbol->isStartPC())
-            {
-            // use inSnippet, as the relocation mechanism is already set up there
-            TR::Register *reg = _baseRegister = cg->allocateRegister();
-            loadAddressConstantInSnippet(cg, nodeForSymbol, 0, reg, NULL, TR::InstOpCode::addi, false, NULL);
-            return;
-            }
-         else if (isStaticField && !ref->isUnresolved())
-            {
-            TR::Register *reg = _baseRegister = cg->allocateRegister();
-            loadAddressConstant(cg, nodeForSymbol, 1, reg, NULL, false, TR_DataAddress);
-            return;
-            }
-         else if (ref->isUnresolved() || useUnresSnippetToAvoidRelo)
-            {
-            self()->setUnresolvedSnippet(new (cg->trHeapMemory()) TR::UnresolvedDataSnippet(cg, node, ref, isStore, false));
-            cg->addSnippet(self()->getUnresolvedSnippet());
-            return;
-            }
+      if ((refIsUnresolved || comp->compileRelocatableCode()) && symbol->isDebugCounter())
+         {
+         TR::Register *reg = _baseRegister = cg->allocateRegister();
+         loadAddressConstant(cg, comp->compileRelocatableCode(), nodeForSymbol, 1, reg, NULL, false, TR_DebugCounter);
+         return;
+         }
+      else if ((refIsUnresolved || cg->needRelocationsForPersistentInfoData() || cg->needRelocationsForStatics()) && symbol->isCountForRecompile())
+         {
+         TR::Register *reg = _baseRegister = cg->allocateRegister();
+         loadAddressConstant(cg, cg->needRelocationsForPersistentInfoData() || cg->needRelocationsForStatics(), nodeForSymbol, TR_CountForRecompile, reg, NULL, false, TR_GlobalValue);
+         return;
+         }
+      else if ((refIsUnresolved || cg->needRelocationsForBodyInfoData() || cg->needRelocationsForStatics()) && symbol->isRecompilationCounter())
+         {
+         TR::Register *reg = _baseRegister = cg->allocateRegister();
+         loadAddressConstant(cg, cg->needRelocationsForBodyInfoData() || cg->needRelocationsForStatics(), nodeForSymbol, 0, reg, NULL, false, TR_BodyInfoAddressLoad);
+         return;
+         }
+      else if (symbol->isCompiledMethod() && (ref->isUnresolved() || comp->compileRelocatableCode() || comp->isOutOfProcessCompilation()))
```

Comment:
Not entirely sure what this part is supposed to be handling, it might be OpenJ9-specific and loading a `J9Method` pointer in preparation for calling the method through the interpreter.

Did we enable the `TR_MethodCallAddress` relocation to be able to do direct calls in JITServer mode on Power or not? If not then I think this case might be one that could be handled by that. I say that because the check is for a compiled method and an unresolved reference or `compileRelocatableCode`, which is a typical pattern for making calls relocatable by using the same code path as unresolved calls.

Reply:
From our Slack conversation, I'm able to recover some notes about this issue. We never actually saw a bug related to this, but while investigating a similar bug, we noticed that TOC code was being generated and we were not entering the `if` here even though we should have been. Initially I was using the `needsRelocationsForStatics()` query here, like `x86` does:

The tree in question looks like this:
```
------------------------------
 n56n     (  0)  treetop                                                                              [    0x7fffa4ce0140] bci=[-1,0,2375] rc=0 vc=194 vn=- li=10 udi=- nc=1
 n52n     (  0)    icall  jitRetranslateCallerWithPreparation[#69  helper Method] [flags 0x400 0x0 ] (in GPR_0067) ()  [    0x7fffa4ce0000] bci=[-1,0,2375] rc=0 vc=194 vn=- li=10 udi=65184 nc=3 flg=0x20
 n53n     (  0)      iconst 0x80000 (in GPR_0069) (X!=0 X>=0 )                                        [    0x7fffa4ce0050] bci=[-1,0,2375] rc=0 vc=194 vn=- li=10 udi=608 nc=0 flg=0x104
 n54n     (  0)      loadaddr  <start-PC>[#323  Static] [flags 0x4000303 0x40 ] (in GPR_0068)         [    0x7fffa4ce00a0] bci=[-1,0,2375] rc=0 vc=194 vn=- li=10 udi=64 nc=0
 n55n     (  0)      loadaddr  <J9Method>[#324  Static] [flags 0x8000307 0x40 ] (in GPR_0066)         [    0x7fffa4ce00f0] bci=[-1,0,2375] rc=0 vc=194 vn=- li=10 udi=64896 nc=0
```

The problem is that `jitRetranslateCallerWithPreparation` causes a recompile from jitted code. We end up in this memory reference routine here via the `loadAddr` evaluator.

Previously the check was `else if (symbol->isCompiledMethod() && (ref->isUnresolved() || cg->needRelocationsForStatics()))`

Based on the above, we wouldn't enter this `if` and end up generating incorrect TOC code at the bottom of this routine (fallback option).

Reply:
OK, it sounds like the `needRelocationsForCurrentMethodPC()` makes more sense here than the `needRelocationsForStatics()` query.

Reply:
Fixed here: c80bd8ce4